lib/file-source/src/file_server.rs

```diff
@@ -202,13 +232,18 @@ impl FileServer {
                         break;
                     }
                     if bytes_read > self.max_read_bytes {
+                        maxed_out_reading_single_file = true;
                         break;
                     }
                 }
                 if bytes_read > 0 {
                     global_bytes_read = global_bytes_read.saturating_add(bytes_read);
                     checkpointer.set_checkpoint(file_id, watcher.get_file_position());
                 }
+                // Do not move on to newer files if we are behind on an older file
+                if self.oldest_first && maxed_out_reading_single_file {
+                    break;
```

Comment:
I'm curious if there is a cleaner way to express this double loop its starting to get a bit hard to follow :(

Reply:
There are some general refactoring I've been thinking about and I might take a quick stab at those after getting this merged.

Reply:
👍 sounds good to me